keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt

```diff
@@ -0,0 +1,120 @@
+  private val unhealthyVetoRepository: UnhealthyVetoRepository,
+  @Value("\${veto.unhealthy.ignore-duration:PT5M}")
+  private val ignoreDuration: String,
+  private val clock: Clock,
+  private val publisher: ApplicationEventPublisher,
+  @Value("\${spinnaker.baseUrl}") private val spinnakerBaseUrl: String,
+  private val diffFingerprintRepository: DiffFingerprintRepository
+) : Veto {
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  override suspend fun check(resource: Resource<*>): VetoResponse {
+    val resourceId = resource.id
+    val hasDiff = diffFingerprintRepository.diffCount(resourceId) > 0
+
+    if (unhealthyVetoRepository.isHealthy(resourceId) || hasDiff) {
+      // we want keel to take action if there's a diff
+      return allowedResponse()
+    }
+
+    val lastAllowedTime = unhealthyVetoRepository.getLastALlowedTime(resourceId) ?: return allowedResponse()
+    val now = clock.instant()
+    val shouldAllow = now.minus(Duration.parse(ignoreDuration)) > lastAllowedTime
+
+    return if (shouldAllow) {
+      log.debug("Resource $resourceId has been ignored for ${friendlyTime(ignoreDuration)}, allowing a recheck.")
+      unhealthyVetoRepository.markAllowed(resourceId)
+      allowedResponse()
+    } else {
+      publisher.publishEvent(ResourceNotificationEvent(resourceId, UNHEALTHY, message(resource)))
+      deniedResponse(message = "Resource is unhealthy with no diff, waiting ${friendlyTime(ignoreDuration)} before rechecking.", vetoArtifact = false)
+    }
+  }
+
+  /**
+   *  Assumption: health is only for clusters, and we have specific requirements
+   *  about what the spec looks like in order to construct the notification link
+   */
+  private fun message(resource: Resource<*>): NotifierMessage {
+    val spec = resource.spec
+    if (spec !is Monikered) {
+      throw UnsupportedResourceTypeException("Resource kind ${resource.kind} must be monikered to construct resource links")
+    }
+    if (spec !is Locatable<*>) {
+      throw UnsupportedResourceTypeException("Resource kind ${resource.kind} must be locatable to construct resource links")
+    }
+    val locations = spec.locations
+    if (locations !is AccountAwareLocations) {
+      throw UnsupportedResourceTypeException("Resource kind ${resource.kind} must be have account aware locations to construct resource links")
+    }
+
+    val params = ClusterViewParams(
+      acct = locations.account,
+      q = resource.spec.displayName,
+      stack = spec.moniker.stack ?: "(none)",
+      detail = spec.moniker.detail ?: "(none)"
+    )
+    val resourceUrl = "$spinnakerBaseUrl/#/applications/${resource.application}/clusters?${params.toURL()}"
+    return NotifierMessage(
+      subject = "${resource.spec.displayName} is unhealthy",
+      body = "<$resourceUrl|${resource.id}> is unhealthy and there is not a diff. " +
+        "Spinnaker cannot correct this problem automatically. " +
+        "Please take manual action to fix it."
+    )
```

Comment:
few detail-oriented thoughts:

1. when we're sending this notification do we have awareness of how long things have been unhealthy? if so, I'm wondering if we could re-word it to something like "<resource> has been unhealthy for 2 minutes...[etc.]"
2. I think the definite tone here ("Spinnaker cannot correct", "Please take manual action") is reassuring, but may not always be completely accurate. Giving a sense of total confidence has the potential to send users into "action mode" when they may need to investigate more before deciding to take action. Is there a way to reword this to still be confidence-boosting without suggesting we know exactly what needs to be done? for example, here's the language we use for the unhappy status:
> Spinnaker has been trying to correct a difference, but taking action hasn't helped. Manual intervention might be required. You can click History to see more.

3. "and there is not a diff" is slightly technical to me — is that detail useful/actionable in this notification?

Reply:
Let me try to address this feedback. I think having a link to the history would be super useful, and I think I can give some sense of timing. I think linking to the history could provide the "no diff" message (what I'm trying to convey is "we have your resource right where you want it, pal, and something is not right)

Reply:
Agree with "there is not a diff" being too technical. I suspect a lot of users won't understand what that means. How about something like this: 
```suggestion
      subject = "${resource.spec.displayName} is unhealthy",
      body = "Spinnaker has detected that <$resourceUrl|${resource.id}> matches its desired state, but is currently unhealthy. " +
        "We cannot correct this problem automatically. Manual intervention might be required. " +
        "You can view the resource history for more information."
    )
```

Reply:
![Screen Shot 2020-09-30 at 5 12 23 PM](https://user-images.githubusercontent.com/8454927/94752591-863f0780-0340-11eb-940b-ad1aa29a1110.png)

Updated to this, what do you two think?